File: python/gemini_agent.py
#!/usr/bin/env python3
"""Gemini agent (your GEMINI_API_KEY) with a PreTestAds score_ad tool.

Uses the google-genai SDK's automatic function calling: pass the Python
function directly and Gemini calls it when needed. The tool pays $5.00 USDC
via x402 (SOLANA_PRIVATE_KEY or EVM_PRIVATE_KEY) and blocks ~5-8 min.

Usage:
    pip install google-genai
    GEMINI_API_KEY=... SOLANA_PRIVATE_KEY=<b58> \
      python gemini_agent.py "Score this ad and suggest improvements: https://example.com/ad.mp4"
"""
import os
import sys

from google import genai
from google.genai import types

import pretestads

MODEL = os.environ.get("GEMINI_MODEL", "gemini-2.5-pro")


def score_ad(media_url: str, window_seconds: int = 15) -> dict:
    """Score a video or image ad with PreTestAds' fMRI-trained neural model.

    Costs $5.00 USDC (paid automatically via x402) and takes 5-8 minutes.
    Returns a 0-100 percentile score vs 76 top TikTok ads, a weak/moderate/strong
    label, hook strength, and a per-second engagement timeseries.

    Args:
        media_url: Public https URL of the ad (mp4/mov/avi/jpg/png/webp).
        window_seconds: Scoring window: 3, 5, 15, 30, or 60 (default 15).
    """
    print(f"[tool] score_ad({media_url}) — paying $5 USDC via x402, then waiting for the model...")
    return pretestads.score_ad(media_url, window_seconds)


def main():
    prompt = " ".join(sys.argv[1:]) or "Score this ad: https://example.com/ad.mp4"
    client = genai.Client()  # reads GEMINI_API_KEY
    resp = client.models.generate_content(
        model=MODEL,
        contents=prompt,
        config=types.GenerateContentConfig(
            system_instruction="You are an ad-creative analyst. Use the score_ad tool when asked to evaluate an ad, then interpret the results with concrete advice.",
            tools=[score_ad],  # automatic function calling
        ),
    )
    print(resp.text)


if __name__ == "__main__":
    main()

File: javascript/gemini-agent.mjs
#!/usr/bin/env node
/**
 * Gemini agent (your GEMINI_API_KEY) with a PreTestAds score_ad tool.
 *
 * Usage:
 *   npm install @google/genai
 *   GEMINI_API_KEY=... EVM_PRIVATE_KEY=0x... \
 *     node gemini-agent.mjs "Score this ad and suggest improvements: https://example.com/ad.mp4"
 */
import { GoogleGenAI, Type } from "@google/genai";
import { scoreAd } from "./lib/pretestads.mjs";

const MODEL = process.env.GEMINI_MODEL ?? "gemini-2.5-pro";

const scoreAdDecl = {
  name: "score_ad",
  description:
    "Score a video or image ad with PreTestAds' fMRI-trained neural model. " +
    "Costs $5.00 USDC (paid automatically via x402) and takes 5-8 minutes. " +
    "Returns a 0-100 percentile score vs 76 top TikTok ads, a weak/moderate/strong label, " +
    "hook strength, and a per-second engagement timeseries.",
  parameters: {
    type: Type.OBJECT,
    properties: {
      media_url: { type: Type.STRING, description: "Public https URL of the ad (mp4/mov/avi/jpg/png/webp)" },
      window_seconds: { type: Type.INTEGER, description: "Scoring window: 3, 5, 15, 30, or 60 (default 15)" },
    },
    required: ["media_url"],
  },
};

const ai = new GoogleGenAI({}); // reads GEMINI_API_KEY
const contents = [
  { role: "user", parts: [{ text: process.argv.slice(2).join(" ") || "Score this ad: https://example.com/ad.mp4" }] },
];
const config = {
  systemInstruction: "You are an ad-creative analyst. Use the score_ad tool when asked to evaluate an ad, then interpret the results with concrete advice.",
  tools: [{ functionDeclarations: [scoreAdDecl] }],
};

for (;;) {
  const resp = await ai.models.generateContent({ model: MODEL, contents, config });
  const calls = resp.functionCalls ?? [];
  if (!calls.length) {
    console.log(resp.text);
    break;
  }
  contents.push(resp.candidates[0].content);
  for (const call of calls) {
    console.log(`[tool] score_ad(${call.args.media_url}) — paying $5 USDC via x402, then waiting...`);
    let result;
    try {
      result = await scoreAd(call.args.media_url, call.args.window_seconds ?? 15);
    } catch (e) {
      result = { error: String(e) };
    }
    contents.push({ role: "user", parts: [{ functionResponse: { name: call.name, response: { result } } }] });
  }
}
